### ISSUE TITLE
Loss turns to NAN suddenly with FP16+AdamW

### ISSUE THREAD

**Issue by baibaidj** [open-mmlab/mmsegmentation#67]
I'm training an unet for a one-class segmentation task using a medical dataset of 10K+ images.
During training, the AdamW optimizer was used for quicker convergence with FP16. The loss function is the weighted sum of BCE and dice-loss with 1.0 and 0.5 coefficients. 
For the first half of the training, everything looked fine.  Then suddenly the loss turns to NAN, like the following log snippet. 

> 2020-08-13 22:55:06,554 - mmseg - INFO - Iter [11264/35238]	lr: 7.217e-04, mIoU: 0.6905, mAcc: 0.7772, aAcc: 0.9943, mDice: 0.7553
> 2020-08-13 22:55:30,333 - mmseg - INFO - Iter [11300/35238]	lr: 7.208e-04, eta: 5:24:59, time: 1.966, data_time: 1.318, memory: 5422, decode.loss_seg: 0.1299, decode.acc_seg: 97.8018, loss: 0.1299
> 2020-08-13 22:55:55,553 - mmseg - INFO - Iter [11350/35238]	lr: 7.196e-04, eta: 5:23:46, time: 0.505, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1319, decode.acc_seg: 97.8515, loss: 0.1319
> 2020-08-13 22:56:20,822 - mmseg - INFO - Iter [11400/35238]	lr: 7.183e-04, eta: 5:22:33, time: 0.505, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1351, decode.acc_seg: 97.8480, loss: 0.1351
> 2020-08-13 22:56:52,062 - mmseg - INFO - Iter [11450/35238]	lr: 7.170e-04, eta: 5:21:33, time: 0.625, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1291, decode.acc_seg: 97.8107, loss: 0.1291
> 2020-08-13 22:57:17,234 - mmseg - INFO - Iter [11500/35238]	lr: 7.158e-04, eta: 5:20:20, time: 0.503, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1296, decode.acc_seg: 97.8579, loss: 0.1296
> 2020-08-13 22:57:42,480 - mmseg - INFO - Iter [11550/35238]	lr: 7.145e-04, eta: 5:19:09, time: 0.505, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1306, decode.acc_seg: 97.8200, loss: 0.1306
> 2020-08-13 22:58:07,794 - mmseg - INFO - Iter [11600/35238]	lr: 7.133e-04, eta: 5:17:57, time: 0.506, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1314, decode.acc_seg: 97.8134, loss: 0.1314
> 2020-08-13 22:58:38,951 - mmseg - INFO - Iter [11650/35238]	lr: 7.120e-04, eta: 5:16:58, time: 0.623, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1318, decode.acc_seg: 97.7512, loss: 0.1318
> 2020-08-13 22:59:04,141 - mmseg - INFO - Iter [11700/35238]	lr: 7.107e-04, eta: 5:15:48, time: 0.504, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1347, decode.acc_seg: 97.8318, loss: 0.1347
> 2020-08-13 22:59:29,405 - mmseg - INFO - Iter [11750/35238]	lr: 7.095e-04, eta: 5:14:38, time: 0.505, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1291, decode.acc_seg: 97.9076, loss: 0.1291
> 2020-08-13 23:00:00,575 - mmseg - INFO - Iter [11800/35238]	lr: 7.082e-04, eta: 5:13:39, time: 0.623, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1274, decode.acc_seg: 97.7749, loss: 0.1274
> 2020-08-13 23:00:25,718 - mmseg - INFO - Iter [11850/35238]	lr: 7.069e-04, eta: 5:12:30, time: 0.503, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1324, decode.acc_seg: 97.8026, loss: 0.1324
> 2020-08-13 23:00:51,036 - mmseg - INFO - Iter [11900/35238]	lr: 7.057e-04, eta: 5:11:21, time: 0.506, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1321, decode.acc_seg: 97.8912, loss: 0.1321
> 2020-08-13 23:01:16,244 - mmseg - INFO - Iter [11950/35238]	lr: 7.044e-04, eta: 5:10:12, time: 0.504, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1269, decode.acc_seg: 97.7969, loss: 0.1269
> 2020-08-13 23:02:12,590 - mmseg - INFO - per class results:
> Class                  IoU        Acc       Dice
> target-class            67.54      85.13      75.49
> Summary:
> Scope                 mIoU       mAcc       aAcc      mDice
> global               67.54      85.13      99.33      75.49
> 
> 2020-08-13 23:02:12,629 - mmseg - INFO - Iter [11968/35238]	lr: 7.040e-04, mIoU: 0.6754, mAcc: 0.8513, aAcc: 0.9933, mDice: 0.7549
> 2020-08-13 23:02:34,288 - mmseg - INFO - Exp name: unet_r34-d32_5bridge_decx1.5_NL000_att0_449x449_200ep_pneumonia_t4v10_fp16.py
> 2020-08-13 23:02:34,288 - mmseg - INFO - Iter [12000/35238]	lr: 7.031e-04, eta: 5:11:43, time: 2.156, data_time: 1.492, memory: 5422, decode.loss_seg: 0.1338, decode.acc_seg: 97.9035, loss: 0.1338
> 2020-08-13 23:02:59,462 - mmseg - INFO - Iter [12050/35238]	lr: 7.019e-04, eta: 5:10:34, time: 0.504, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1297, decode.acc_seg: 97.8303, loss: 0.1297
> 2020-08-13 23:03:24,521 - mmseg - INFO - Iter [12100/35238]	lr: 7.006e-04, eta: 5:09:25, time: 0.501, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1319, decode.acc_seg: 97.7559, loss: 0.1319
> 2020-08-13 23:03:55,583 - mmseg - INFO - Iter [12150/35238]	lr: 6.993e-04, eta: 5:08:27, time: 0.621, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1323, decode.acc_seg: 97.8359, loss: 0.1323
> 2020-08-13 23:04:20,744 - mmseg - INFO - Iter [12200/35238]	lr: 6.981e-04, eta: 5:07:19, time: 0.503, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1317, decode.acc_seg: 97.6793, loss: 0.1317
> 2020-08-13 23:04:45,974 - mmseg - INFO - Iter [12250/35238]	lr: 6.968e-04, eta: 5:06:11, time: 0.505, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1338, decode.acc_seg: 97.8534, loss: 0.1338
> 2020-08-13 23:05:11,236 - mmseg - INFO - Iter [12300/35238]	lr: 6.956e-04, eta: 5:05:04, time: 0.505, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1258, decode.acc_seg: 97.9049, loss: 0.1258
> 2020-08-13 23:05:42,341 - mmseg - INFO - Iter [12350/35238]	lr: 6.943e-04, eta: 5:04:08, time: 0.622, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1279, decode.acc_seg: 97.8606, loss: 0.1279
> 2020-08-13 23:06:07,630 - mmseg - INFO - Iter [12400/35238]	lr: 6.930e-04, eta: 5:03:01, time: 0.506, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1298, decode.acc_seg: 97.8347, loss: 0.1298
> 2020-08-13 23:06:32,892 - mmseg - INFO - Iter [12450/35238]	lr: 6.918e-04, eta: 5:01:55, time: 0.505, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1271, decode.acc_seg: 97.8194, loss: 0.1271
> 2020-08-13 23:07:04,092 - mmseg - INFO - Iter [12500/35238]	lr: 6.905e-04, eta: 5:00:59, time: 0.624, data_time: 0.012, memory: 5422, decode.loss_seg: 0.1293, decode.acc_seg: 97.8613, loss: 0.1293
> 2020-08-13 23:07:29,326 - mmseg - INFO - Iter [12550/35238]	lr: 6.892e-04, eta: 4:59:54, time: 0.505, data_time: 0.013, memory: 5422, decode.loss_seg: 0.1281, decode.acc_seg: 97.8741, loss: 0.1281
> 2020-08-13 23:07:54,609 - mmseg - INFO - Iter [12600/35238]	lr: 6.879e-04, eta: 4:58:48, time: 0.506, data_time: 0.013, memory: 5422, decode.loss_seg: nan, decode.acc_seg: 97.7602, loss: nan
> 2020-08-13 23:08:19,955 - mmseg - INFO - Iter [12650/35238]	lr: 6.867e-04, eta: 4:57:43, time: 0.507, data_time: 0.013, memory: 5422, decode.loss_seg: nan, decode.acc_seg: 97.8386, loss: nan
> 2020-08-13 23:09:18,275 - mmseg - INFO - per class results:
> Class                  IoU        Acc       Dice
> target-class             0.00       0.00       5.56
> Summary:
> Scope                 mIoU       mAcc       aAcc      mDice
> global                0.00       0.00      98.37       5.56
> 
> 2020-08-13 23:09:18,313 - mmseg - INFO - Iter [12672/35238]	lr: 6.861e-04, mIoU: 0.0000, mAcc: 0.0000, aAcc: 0.9837, mDice: 0.0556

In the optimizer config, I even used gradient clipping like follows. 

```
optimizer = dict(
    type='AdamW',
    lr=0.001,
    weight_decay=0.0005,
    paramwise_cfg=dict(custom_keys=dict(head=dict(lr_mult=4.0))))
optimizer_config = dict(
    type='Fp16OptimizerHook',
    loss_scale=512.0,
    grad_clip=dict(max_norm=10, norm_type=2))
lr_config = dict(policy='poly', power=0.9, min_lr=5e-05, by_epoch=False)
```

Hi. in another related thread, @xvjiarui  gave two suggestions, which I do not think quite apply for my case. 1. warmup does not apply because my model converged fine for 10K iters. 2.  The current lr was quite low in my settings although I can try lower. But what if that doesn't work?

Thanks.

**Comment by xvjiarui** [comment]
Hi @baibaidj 
I haven't tried AdamW and dice-loss. I think AdamW should be fine given the previous iteration results (you may also check some papers with AdamW if they used poly decay). But I am not sure if dice loss would cause NaN. 
I did experiments with SGD and CrossEntropy, for most models it works fine. 
I suggest you run you 2 experiments with `--seed 0` option to reduce randomness, one with dice-loss, other not, to see if both experiments would run into NaN.

**Comment by baibaidj** [comment]
> Hi @baibaidj
> I haven't tried AdamW and dice-loss. I think AdamW should be fine given the previous iteration results (you may also check some papers with AdamW if they used poly decay). But I am not sure if dice loss would cause NaN.
> I did experiments with SGD and CrossEntropy, for most models it works fine.
> I suggest you run you 2 experiments with `--seed 0` option to reduce randomness, one with dice-loss, other not, to see if both experiments would run into NaN.

Thanks. After I decreased the lr and removed the lr_mult for decoder head, the loss-Nan issue went away.
